library/ziti-socket.c

```diff
@@ -0,0 +1,404 @@
+    if (status == ZITI_OK) {
+        ziti_conn_bridge_fds(conn, (uv_os_fd_t) zs->ziti_fd, (uv_os_fd_t) zs->ziti_fd, on_bridge_close, zs);
+        complete_future(zs->f, conn);
+    } else {
+        fail_future(zs->f, status);
+        ziti_close(zs->conn, NULL);
+        on_bridge_close(zs);
+    }
+}
+
+static void do_ziti_connect(struct dial_req_s *req, future_t *f, uv_loop_t *l) {
+    ziti_sock_t *zs = model_map_get_key(&ziti_sockets, &req->fd, sizeof(req->fd));
+    if (zs == NULL) {
+        ZITI_LOG(WARN, "socket %ld not found", req->fd);
+        fail_future(f, -EBADF);
+    } else if (zs->f != NULL) {
+        fail_future(f, -EALREADY);
+    } else {
+        zs->ztx = req->ztx;
+        zs->f = f;
+        ziti_conn_init(req->ztx, &zs->conn, zs);
+        ziti_dial(zs->conn, req->service, on_ziti_connect, NULL);
+    }
+}
+
+int Ziti_connect(ziti_socket_t socket, ziti_context ztx, const char *service) {
+    struct dial_req_s req = {
+            .fd = socket,
+            .ztx = ztx,
+            .service = service
+    };
+    future_t *f = schedule_on_loop((loop_work_cb) do_ziti_connect, &req, true);
+    int err = await_future(f);
+    destroy_future(f);
+    return err;
+}
+
+void Ziti_lib_shutdown(void) {
+    schedule_on_loop(do_shutdown, NULL, true);
+    uv_thread_join(&lib_thread);
+#if _WIN32
+    closesocket(ziti_sock_server);
+    if (!DeleteFile(ziti_sock_name.sun_path)) {
+        fprintf(stderr, "failed to delete file: %lu\n", GetLastError());
+    }
+#endif
+}
+
+static void looper(void *arg) {
+    uv_run(arg, UV_RUN_DEFAULT);
+}
+
+future_t *schedule_on_loop(loop_work_cb cb, const void *arg, bool wait) {
+    queue_elem_t *el = calloc(1, sizeof(queue_elem_t));
+    el->cb = cb;
+    el->arg = arg;
+
+    uv_mutex_lock(&q_mut);
+    LIST_INSERT_HEAD(&loop_q, el, _next);
+    uv_mutex_unlock(&q_mut);
+    uv_async_send(&q_async);
+
+    if (wait) {
+        el->f = new_future();
+    }
+    return el->f;
+}
+
+void process_on_loop(uv_async_t *async) {
+    LIST_HEAD(loop_queue, queue_elem_s) q = {0};
+
+    // drain q
+    uv_mutex_lock(&q_mut);
+    while (!LIST_EMPTY(&loop_q)) {
+        queue_elem_t *el = LIST_FIRST(&loop_q);
+        LIST_REMOVE(el, _next);
+        LIST_INSERT_HEAD(&q, el, _next);
+    }
+    uv_mutex_unlock(&q_mut);
+
+    while (!LIST_EMPTY(&q)) {
+        queue_elem_t *el = LIST_FIRST(&q);
+        LIST_REMOVE(el, _next);
+        el->cb(el->arg, el->f, async->loop);
+        free(el);
+    }
+}
+
+static void internal_init() {
+#if _WIN32
+    WORD ver = MAKEWORD(2,2);
+    WSADATA data;
+    DWORD err;
+    err = WSAStartup(ver, &data);
+
+    ziti_sock_server = socket(AF_UNIX, SOCK_STREAM, 0);
+    if (ziti_sock_server == INVALID_SOCKET) {
+        fprintf(stderr, "invalid socket: %d", WSAGetLastError());
+    }
+
+    ziti_sock_name.sun_family = AF_UNIX;
+    char temp[sizeof(ziti_sock_name.sun_path)];
+    GetTempPath(sizeof(temp), temp);
```

Comment:
this may return a path that is writable by users other than the caller, and the ownership/permissions of this file determine the permissions of the socket (from [AF_UNIX comes to Windows](https://devblogs.microsoft.com/commandline/af_unix-comes-to-windows/#security)). Maybe `%LOCALAPPDATA%` or explicitly creating a directory with the desired permission would be better?

Reply:
`GetTempPath()` returns `%LOCALAPPDATA%\Temp` which is not accessible by other users (except admin)

Reply:
`GetTempPath` returns `%LOCALAPPDATA%` at least in some cases. [The doc](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-gettemppatha) says it may return other values:

> The GetTempPath function checks for the existence of environment variables in the following order and uses the first path found:
> The path specified by the TMP environment variable.
> The path specified by the TEMP environment variable.
> The path specified by the USERPROFILE environment variable.
> The Windows directory.